packages/fx-core/src/component/resource/botService/botRegistration/localBotRegistration.ts

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { IBotRegistration } from "../appStudio/interfaces/IBotRegistration";
+import { err, FxError, Result, ok, M365TokenProvider } from "@microsoft/teamsfx-api";
+import { AppStudioScopes } from "../../../../common/tools";
+import { AppStudioClient } from "../appStudio/appStudioClient";
+import { BotRegistration, BotAuthType, IBotAadCredentials } from "./botRegistration";
+
+export class LocalBotRegistration extends BotRegistration {
```

Comment:
A bit confused on what's a `BotRegistration`? Is it *AAD + Bot*, or just *AAD*?
Then what's will be suggested way to call `LocalBotRegistration` from `debugHandler/bot.ts`?

IMO, if `BotRegistration` represents *AAD + Bot*, `debugHandler` can direct call `LocalBotRegistration.create...`
Otherwise, if `BotRegistration` represents *AAD*, `debugHandler` needs to call `LocalBotRegistration..` first, then `AppStudioClient...`, and can get rid of `GraphClient`...

Correct me if I miss understand anything : )

Reply:
`BotRegistration` is AAD (or MSI in the future) + Bot Registration (By AppStudio or Azure Bot Service). In a word, it's a wrapper around bot registrations no matter in which way it's implemented.

The first way you mentioned above is the right way in my mind. The wrapper here is trying to make bot registrations consistent across local and remote. The ideal way to call `BotRegistration` is like below:
```
const botRegistration: BotRegistration = new AppStudioBotRegistration(); // aka LocalBotRegistration
botRegistration.createBotRegistration(); // Create an AAD app and also do bot registration.
botRegistration.updateMessageEndpoint();
// or
const botRegistration: BotRegistration = new AzureBotRegistration(); // aka RemoteBotRegistration
botRegistration.createBotRegistration(); // Just create an AAD app for bot.
```
I don't know if we will support registering bot by Azure Bot in local debug, but the current design is flexible enough to adapt that change.

I'm wondering if renaming LocalBotRegistration to AppStudioBotRegistration and RemoteBotRegistration to AzureBotRegistration are better. Just in considering this.

Reply:
@swatDong BTW, I'm not familiar enough to change the code of local debug to adapt my current design, so didn't do too many changes. Can your team help use the `BotRegistration` in local debug? Thanks.

Reply:
Sure, the local debug part can be updated later. Just to confirm:
1. Are `GraphClient` and `AppStudioClient` going to be retired?
2. Still confuse on the naming. As you said, "`BotRegistration` is AAD (or MSI in the future) + Bot Registration (By AppStudio or Azure Bot Service).", why `AzureBotRegistration.createBotRegistration()` only create AAD (or MSI)?